### PR TITLE
Add tests for casting UNKNOWN type to any other type

### PR DESCRIPTION
**Pull request by PHILO-HE**

### CODE REVIEW

**Comment by PHILO-HE** [comment]
@mbasmanova, could you spare some time to review this patch? Thanks!

File: velox/exec/tests/utils/QueryAssertions.cpp [hunk 1]
```diff
@@ -274,6 +274,14 @@ variant variantAt<TypeKind::TIMESTAMP>(
       dataChunk->GetValue(column, row).GetValue<::duckdb::timestamp_t>()));
 }
 
+template <>
+variant variantAt<TypeKind::UNKNOWN>(
```
Comment (mbasmanova) [inline velox/exec/tests/utils/QueryAssertions.cpp:278]: When would this be called? I believe there is a check for null before this call. If that's the case, then, perhaps, a safer implementation would be to throw VELOX_UNREACHABLE.

File: velox/expression/CastExpr.cpp [hunk 2]
```diff
@@ -724,6 +734,9 @@ void CastExpr::applyPeeled(
             fromType->asRow(),
             toType);
         break;
+      case TypeKind::UNKNOWN:
+        result = applyUnknown(rows, input, context, fromType, toType);
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:738]: Why not just inline

```
result = BaseVector::createNullConstant(toType, rows.end(), context.pool());
```

File: velox/expression/CastExpr.h [hunk 3]
```diff
@@ -113,6 +113,13 @@ class CastExpr : public SpecialForm {
       const TypePtr& toType,
       VectorPtr& result);
 
+  VectorPtr applyUnknown(
```
Comment (mbasmanova) [inline velox/expression/CastExpr.h:116]: This method is not using any member variables, hence, can be a free function in .cpp file. Also, the implementation is a one-liner and there is only one caller, hence, this function is not needed at all.

File: velox/expression/tests/CastExprTest.cpp [hunk 4]
```diff
@@ -371,6 +371,19 @@ TEST_F(CastExprTest, basics) {
       {"1.888", "2.5", "3.6", "100.44", "-100.101", "1", "-2"});
 }
 
+TEST_F(CastExprTest, fromUnknownType) {
+  testCast<UnknownValue, int>(
```
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:375]: int -> int32_t

Can we test all types?

File: velox/functions/prestosql/aggregates/PrestoHasher.cpp [hunk 5]
```diff
@@ -116,6 +116,13 @@ FOLLY_ALWAYS_INLINE void PrestoHasher::hash(
   hashIntegral<T>(*vector_.get(), rows, hashes);
 }
 
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::UNKNOWN>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(rows, *vector_.get(), hashes, [&](auto row) { return 0; });
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/PrestoHasher.cpp:123]: Would you add a test case for this change?

File: velox/functions/prestosql/types/JsonType.cpp [hunk 6]
```diff
@@ -586,14 +586,21 @@ simdjson::error_code appendMapKey(
...
     SIMDJSON_ASSIGN_OR_RAISE(writer.castTo<T>(), fromString<T>(value));
     return simdjson::SUCCESS;
   }
 }
 
+template <>
+simdjson::error_code appendMapKey<TypeKind::UNKNOWN>(
+    const std::string_view& value,
+    exec::GenericWriter& writer) {
+  VELOX_NYI("UNKNOWN type is not supported!");
```
Comment (mbasmanova) [inline velox/functions/prestosql/types/JsonType.cpp:601]: Will this ever be called? Isn't there a null check before this call? If so, VELOX_UNREACHABLE would be more appropriate.

File: velox/type/Type.h [hunk 7]
```diff
@@ -129,6 +129,10 @@ struct UnknownValue {
   bool operator>=(const UnknownValue& /* b */) const {
     return true;
   }
+
+  operator std::string() const {
+    return "NULL";
```
Comment (mbasmanova) [inline velox/type/Type.h:134]: What is this change for? Would it be possible to add a test?

File: velox/type/Type.h [hunk 8]
```diff
@@ -1420,6 +1424,10 @@ std::shared_ptr<const OpaqueType> OPAQUE() {
         return TEMPLATE_FUNC<::facebook::velox::TypeKind::TIMESTAMP>(         \
             __VA_ARGS__);                                                     \
       }                                                                       \
+      case ::facebook::velox::TypeKind::UNKNOWN: {                            \
+        return TEMPLATE_FUNC<::facebook::velox::TypeKind::UNKNOWN>(           \
```
Comment (mbasmanova) [inline velox/type/Type.h:1428]: Do we need this change? It will require a lot of updates to existing code. Would it be possible to avoid it?

**Review by mbasmanova — COMMENTED** [review]
@PHILO-HE Curious, what's the motivation for this PR?

**Comment by PHILO-HE** [comment]
> @PHILO-HE Curious, what's the motivation for this PR?

Hi @mbasmanova, thanks for your review! The motivation is we are trying to fix issues when enabling Spark NullType (UNKNOWN type of Velox).
I just divided this PR and removed the code irrelevant to the topic of this pr. Could you review again?

File: velox/exec/tests/utils/QueryAssertions.cpp (hunk 1 above)
Comment (PHILO-HE) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Just removed it and will cover it in another PR.

File: velox/type/Type.h [hunk 9]
```diff
@@ -2390,6 +2390,16 @@ struct IsRowType<Row<Ts...>> {
 
 } // namespace facebook::velox
 
+namespace std {
+template <>
+struct hash<::facebook::velox::UnknownValue> {
```
Comment (PHILO-HE) [inline velox/type/Type.h:2395]: The hash for UNKNOWN type is required by VectorMakerStats. If lacked, compiling will fail.
See https://github.com/facebookincubator/velox/blob/main/velox/vector/tests/utils/VectorMakerStats.h#L86.

Comment (mbasmanova) [reply velox/type/Type.h:2395]: VectorMakerStats is deprecated. It would be great to not use it (and eventually delete it).

**Review by mbasmanova — APPROVED** [review]
Looks good % let's update documentation.

**Comment by PHILO-HE** [comment]
Hi @mbasmanova, I found the added tests can pass without any change for `CastExpr.cpp`. So I just removed the unused change and changed the PR title.

File: velox/type/Type.h (hunk 9 above)
Comment (PHILO-HE) [reply velox/type/Type.h]: I just removed this and introduced some code in `VectorMakerStats.h` to make the compiling pass. Maybe, better to delete deprecated code in a separate PR.

File: velox/docs/functions/spark/conversion.rst [hunk 10]
```diff
@@ -2,6 +2,8 @@
 Conversion Functions
 ====================
 
+Casting from UNKNOWN type to all other scalar types are supported, e.g., cast(NULL as int).
```
Comment (mbasmanova) [inline velox/docs/functions/spark/conversion.rst:5]: typo: "Casting.. are supported" -> "Casting... is supported"

**Review by mbasmanova — COMMENTED** [review]
@PHILO-HE Looks good % one nit.

File: velox/docs/functions/spark/conversion.rst (hunk 10 above)
Comment (PHILO-HE) [reply velox/docs/functions/spark/conversion.rst]: @mbasmanova, just fixed it and rebased the code. Thanks!

**Pull request closed without merging**